src/simplify_algebra.cpp

```diff
@@ -506,65 +518,124 @@ struct find_inner_broadcast
+                    idims.push_back(ins->get_shape().lens()[axis]);
                 }
             }
-            // if no common broadcast axis, transformation is not useful
-            if(std::find_if(common_axis.begin(), common_axis.end(), [](auto num_common) {
-                   return num_common > 1;
-               }) == common_axis.end())
+            if(iaxes.size() == ndim)
                 return;
+            std::vector<instruction_ref> inputs;
+            std::transform(
+                broadcasts.begin(),
+                broadcasts.end(),
+                std::back_inserter(inputs),
+                [&](instruction_ref broadcast) {
+                    auto input = broadcast->inputs().front();
+                    auto s     = input->get_shape();
+
+                    if(s.elements() == 1)
+                    {
+                        if(s.lens().size() > 1)
+                            return m.insert_instruction(broadcast, make_op("squeeze"), input);
+                        return input;
+                    }
+
+                    std::int64_t shift = ndim - s.ndim();
+                    if(broadcast->name() == "broadcast")
+                        shift = broadcast->get_operator().to_value()["axis"].to<int64_t>();
+                    std::vector<std::int64_t> sq_axes;
+                    for(auto axis : axes)
+                    {
+                        auto iaxis = axis - shift;
+                        if(iaxis < 0)
+                            continue;
+                        if(iaxis >= s.ndim())
+                            continue;
+                        sq_axes.push_back(iaxis);
+                    }
+                    instruction_ref result = input;
+                    if(not sq_axes.empty())
+                        result = m.insert_instruction(
+                            broadcast, make_op("squeeze", {{"axes", sq_axes}}), result);
+                    if(result->get_shape().ndim() < iaxes.size())
+                    {
+                        auto start_axis = std::find_if(iaxes.begin(),
+                                                       iaxes.end(),
+                                                       [&](auto x) { return x >= shift; }) -
+                                          iaxes.begin();
+                        result = m.insert_instruction(
+                            broadcast,
+                            make_op("broadcast", {{"axis", start_axis}, {"out_lens", idims}}),
+                            result);
+                    }
+                    return result;
+                });
+            auto op = insert_common_op(m, ins, ins->get_operator(), inputs);
+            if(iaxes.size() == 1)
+            {
+                m.replace_instruction(
+                    ins,
+                    make_op("broadcast",
+                            {{"axis", iaxes.front()}, {"out_lens", ins->get_shape().lens()}}),
+                    op);
+            }
```

Comment:
Why use `broadcast` rather than `multibroadcast`?
I understand that for one "inner axis" case that `broadcast` can be used directly rather than `multibroadcast` which might need an `unsqueeze` operator. But, this makes the code more complicated for a seemingly unnoticeable benefit. Another option would be to use `broadcast` for everything since the effective difference between the two operators if that broadcast works left-wise with an offset while multibroadcast works right-wise like in numpy.

Reply:
I think its to keep the same semantics of `broadcast` being for one dimension and `multibroadcast` for multiple. There are places that only look for `multibroadcast` so this could be a problem, and using it without a unsqueeze simplifies the graph for other optimizations. 

Furthermore this reduces the amount of test cases that need to be updated.